### ISSUE TITLE
try to add acc-filter to replay buffer, but hangs and timeout

### ISSUE THREAD

**Issue by yyht** [OpenRLHF/OpenRLHF#717]
I tried to implement acc-filter as proposed in PRIME-verl
`
for rand_prompts in self.prompts_dataloader:
                self.strategy.print({
                    'random_prompts': len(rand_prompts),
                    'info': 'RANDOM'
                })
                for i, experience in enumerate(
                    self.experience_maker.make_experience_list(rand_prompts, **self.generate_kwargs)
                ):
                    if i == 0:
                        output = self.tokenizer.batch_decode(
                            experience.sequences[0].unsqueeze(0), skip_special_tokens=False
                        )
                        self.strategy.print(output)
                    self.replay_buffer.append(experience)

                if self.strategy.args.prompt_filter:
                    self.strategy.print({
                        'BUFFERSIZEBEFOREFILTER': len(self.replay_buffer)
                    })
                    self.strategy.print("###doing_prompt_filter###")
                    self.replay_buffer.filter_items()
                    self.strategy.print({
                        'BUFFERSIZEAFTERFILTER': len(self.replay_buffer)
                    })
                    if len(self.replay_buffer) < self.args.n_samples_per_prompt * len(rand_prompts):
                        continue
                    if len(self.replay_buffer) > self.args.n_samples_per_prompt * len(rand_prompts):
                        self.replay_buffer.items = self.replay_buffer.items[:self.args.n_samples_per_prompt * len(rand_prompts)]

                    self.strategy.print({
                        'BUFFERSIZE': len(self.replay_buffer)
                    })

                self.strategy.print("##PREPARE_TO_TRAIN##")
                torch.distributed.barrier()
                self.strategy.print("##BEGIN_TO_TRAIN##")

                torch.cuda.empty_cache()
                self.replay_buffer.normalize("advantages", self.strategy)
                status = self.ppo_train(steps)
                self.replay_buffer.clear()
                torch.cuda.empty_cache()`

however, it will always hangs on and timeout. 
The replaybuffer seems to be local for each actor worker, and how to sync it between different workers

**Comment by wangzhen0518** [comment]
Hello! I have implemented an `all_gather` method in the `NaiveReplayBuffer` class to synchronize replay buffers across different processes.

```python
import itertools
import torch.distributed as dist

class NaiveReplayBuffer(ABC):
    def __init__(...):
        ...
        self.gathered = False

    ...

    def clear(self) -> None:
        ...
        self.gathered = False

    ...

    def all_gather(self):
        all_items: list[list[BufferItem]] = [None] * dist.get_world_size()  # type: ignore
        dist.all_gather_object(all_items, self.items)
        self.items = [
            item.to_device(self.target_device) if not self.cpu_offload else item
            for item in itertools.chain.from_iterable(all_items)
        ]
        self.gathered = True
```

The `to_device` method of `BufferItem` functions similarly to that of `Experience`.

```python
@dataclass
class BufferItem:
    ...
    @torch.no_grad()
    def to_device(self, device: torch.device):
        self.sequences = to(self.sequences, device)
        self.action_log_probs = to(self.action_log_probs, device)
        self.base_action_log_probs = to(self.base_action_log_probs, device)
        self.values = to(self.values, device)
        self.returns = to(self.returns, device)
        self.advantages = to(self.advantages, device)
        self.attention_mask = to(self.attention_mask, device)
        self.action_mask = to(self.action_mask, device)
        self.info = {key: to(value, device) for key, value in self.info.items()}
        return self
```

The `normalize` method also requires modification because it previously gathered all advantages across different processes. However, since we have already gathered all data by `all_gather`, gathering again would result in duplication. Here is the modified `normalize` method:

```python
class NaiveReplayBuffer(ABC):
    ...
    def normalize(self, attribute: str, strategy) -> None:
        assert attribute == "advantages"
        items = []
        action_masks = []
        for item in self:
            items.append(getattr(item, attribute))
            action_masks.append(item.action_mask)

        items_vector = torch.cat(items).float().flatten()

        if action_masks[0] is None:
            # Samples packed without an action mask
            action_masks_vector = 1
            num_actions = items_vector.numel()
        else:
            action_masks_vector = torch.cat(action_masks).flatten()
            num_actions = action_masks_vector.sum()

        # For Data Parallel (DP)
        # Calculate mean
        sum_and_count = (items_vector.sum(), num_actions)
        all_sum, all_count = (
            strategy.all_reduce(torch.tensor(sum_and_count, device=items_vector.device), "sum")
            if not self.gathered
            else sum_and_count
        )
        mean = all_sum / all_count
        # Calculate standard deviation
        std = ((items_vector - mean).pow(2) * action_masks_vector).sum()
        all_std = strategy.all_reduce(std, "sum") if not self.gathered else std
        rstd = (all_std / all_count).clamp(min=1e-8).rsqrt()

        for i, item in enumerate(self):
            setattr(item, attribute, (items[i] - mean) * rstd)
```

Note that after using `all_gather`, each process's replay buffer contains all training data, so you need to use `DistributedSampler` or `strategy.setup_dataloader`. Here is an example:

```python
# openrlhf/trainer/ppo_trainer.py
class PPOTrainer(ABC):
    def fit(...):
        ...
        for episode in range(start_episode, args.num_episodes):
            ...
            for rand_prompts, labels in self.prompts_dataloader:
                for i, experience in enumerate(
                    self.experience_maker.make_experience_list(rand_prompts, labels, **self.generate_kwargs)
                ):
                    ...
                    self.replay_buffer.append(experience)

                self.replay_buffer.all_gather()
                if self.args.advantage_estimator != "group_norm":
                    self.replay_buffer.normalize("advantages", self.strategy)
                status = self.ppo_train(steps)
                self.replay_buffer.clear()
                ...

    def ppo_train(self, global_steps=0):
        dataloader = self.strategy.setup_dataloader(
            self.replay_buffer,
            self.replay_buffer.sample_batch_size,
            pin_memory=True,
            shuffle=True,
            # drop_last=True, it depends on you
            collate_fn=self.replay_buffer.collate_fn,
        )
        ...
```

**Comment by yyht** [comment]
very nice.

**Comment by LeoDu0314** [comment]
@yyht @wangzhen0518 
hi, I followed the code above, but after filtering, i encountered this problem during training phase

assert not param.ds_active_sub_modules, param.ds_summary()

it seems to be a problem caused by gradient accumulation https://github.com/OpenRLHF/OpenRLHF/issues/630

do you have any ideas how to solve this, thanks a lot!

**Comment by yyht** [comment]
You can use the code from open-reaonser-orz (https://github.com/Open-Reasoner-Zero/Open-Reasoner-Zero/blob/main/orz/ppo/actors.py def ppo_train(self, global_steps, replay_buffer))

Besides, if you want to set different update-steps for actor/critic, you could use the following code:

```
num_replicas = dist.get_world_size() // self.strategy.ring_attn_size
num_samples = math.ceil(len(self.items) / num_replicas)

residual = np.mod(num_samples, max([self.strategy.args.policy_update_steps, 
                                    self.strategy.args.critic_update_steps]))

if residual != 0:
    total_size = num_samples - residual
else:
    total_size = num_samples

self.items = self.items[:(total_size * num_replicas)]
```

This works for 7b model, but for 32b-model, it will arise the another issue.

**Comment by LeoDu0314** [comment]
Thank you very much.
I tried adding this piece of code to the end of `add_gather` and solved my problem.
```
cutoff = len(self.items) % args.train_batch_size
if cutoff != 0:
    self.items = self.items[:-cutoff]
```

don't if it will work for a 32b-model.

**Comment by Unfinito** [comment]
Just a note if one use the pretrain_data and face the same issue: You should cutoff your data using `args.train_batch_size * 2` not `args.train_batch_size`. Because the real `train_batch_size` is doubled here when `self.args.pretrain_data is not None`
https://github.com/OpenRLHF/OpenRLHF/blob/ecbe3b845f917c214cd1407d1b64ebedf1ffb354/openrlhf/utils/deepspeed/deepspeed.py#L238-L239
